cpio and zstd

keszybz opened the pull request:
This adds support for building cpio.zstd archives. See https://github.com/keszybz/mkosi-initrd for instructions how to create a functional pure-systemd initrd from Fedora packages.

I'm marking it as RFC. The cpio part is fine I think. The *implementation* of compression with zstd is also OK. But the user interface needs some design: right now we have `--xz` option, based on the assumption that `xz` is the only compression one would want to use. I think this isn't really true any more: `zstd` achieves very similar compression at a fraction of the compression/decompression times. For images that are widely distributed `xz` is still probably the best choice, but `zstd` is better for local work and anything where speed matters.

So I think we should deprecate the `--xz` option, and have `--inner-compression` (internally in the fs or archive where supported) and `--outer-compression` (for doing compression of the full output, like `--xz` right now), and `--compression` that would pick `--inner-compression` or `--outer-compression` as appropriate. Most of the time users would specify `--compression=yes` or `--compression=zstd|xz|gz|…` and `mkosi` would apply it in the appropriate place.

If you think that's the way to go, I will rework the pull request along those lines.

behrmann commented:
Very cool!

We already hat a PR about adding `--zstd` in addition to `--xz`, but it stalled.

keszybz commented:
That was #465. Yeah, I don't think another `--zstd` switch is the way to go. The "best" compression algorithm changes surprisingly often.

poettering commented:
not a fan of the "inner" and "outer" compression. I'd be precise here, and call one "fs" compression an the other "archive" compression. Naming is hard...

poettering commented on mkosi/__init__.py:
```diff
@@ -5998,12 +5998,13 @@ def load_args(args: argparse.Namespace) -> CommandLineArguments:
+            die("Sorry, can't {opname} with a compressed image.")
         if args.qcow2:
-            die("Sorry, can't acquire shell in or boot a qcow2 image.")
+            die("Sorry, can't {opname} using a qcow2 image.")
```
split out this one into a separate PR and merge right away?

keszybz replied:
It's interleaved with subsequent changes… I had to make a few attempts to have sentence structure which works well with templating of all the options.

poettering commented on mkosi/__init__.py:
```diff
@@ -3721,20 +3722,63 @@ def make_tar(args: CommandLineArguments, root: str, do_run_build_script: bool, f
         return None
 
-    tar_root_dir = f"{root}/usr" if args.usr_only else root
+    root_dir = f"{root}/usr" if args.usr_only else root
```
Maybe rename the function make_tar() → make_archive(). The log message already talks about "archive" anyway...

keszybz replied:
I changed tar→archive in various places to make the code more similar in `make_tar` and `make_cpio`. But this function should still be `make_tar`, because it creates a specific type of archive, and there are other functions for other types.

poettering commented on mkosi/__init__.py:
```diff
@@ -3721,20 +3722,63 @@ def make_tar(args: CommandLineArguments, root: str, do_run_build_script: bool, f
+                stdout=subprocess.PIPE,
+            ) as cpio:
+                with subprocess.Popen(
+                    ["pxz", "--check=crc32", "--lzma2=dict=1MiB", "-T0"], stdin=cpio.stdout, stdout=f
```
share the infra from xz_binary() here to use pxz only if installed, and use plain xz otherwise

keszybz replied:
That chunk is replaced later on with something that does the check…

poettering commented on mkosi/__init__.py:
```diff
@@ -2304,6 +2304,10 @@ def install_fedora(args: CommandLineArguments, root: str, do_run_build_script: b
 
+    # FIXME: should this be conditionalized on args.with_docs like in install_debian_or_ubuntu()?
+    #        But we set LANG=C.UTF-8 anyway.
+    shutil.rmtree(os.path.join(root, "usr/share/locale"))
```
isn't there an official way to get rid of this in fedora?

keszybz replied:
Not directly. `rpm` has `--excludepaths`, but I'm not sure if there's a way to tell dnf to pass that.

keszybz replied:
https://bugzilla.redhat.com/show_bug.cgi?id=1967970

poettering replied:
hmm, maybe we should add a new generic RemoveFiles= stanza that works on all distros, and can be used here? It would just list paths (with wildcards) to remove as very last step of the build or so?

poettering replied:
And then maybe explicit provide an example in the docs, that it's a good idea to list /usr/share/locale in this setting

keszybz replied:
`RemoveFiles=` sounds good. But let's do that separately, this one is already getting rather complex.

poettering commented on mkosi/__init__.py:
```diff
@@ -3756,6 +3756,25 @@ def file_list(root: str) -> Generator[str, None, None]:
                 queue.append(entry.path)
 
 
+def cpio_compressor_command(option: Union[None, str, bool]) -> List[str]:
```
shouldn't we share this with the tar codepaths? i.e. if we invoke the compressor manually for the cpio case there shouldn't be any reason why we couldn't do this for the tar case the exact same way?

keszybz replied:
Yeah, it would be nice to share this. I didn't do that so far because that'll require a bigger restructuring of how `tar` is invoked.

keszybz commented:
> not a fan of the "inner" and "outer" compression. I'd be precise here, and call one "fs" compression an the other "archive" compression. Naming is hard...

So... let's say we add EROFS support. (I think this could be a nicer alternative to squashfs images.)
EROFS has inner compression (lz4 only?). It might be useful to add outer compression with `xz` or `zstd` to shave off some bytes for distribution. It doesn't fit to call the outer compression "archive" in this case, because there's not archive involved.

keszybz commented:
I think this is mostly ready. The last commit is the compression part, significantly reworked along the suggested lines.

keszybz commented:
```
Run python3 -m mypy mkosi
mkosi/__init__.py:3871:21: error: Item "None" of "Optional[IO[Any]]" has no
attribute "write"  [union-attr]
                        cpio.stdin.write(file.encode("utf8") + b"\0")
                        ^
mkosi/__init__.py:3872:17: error: Item "None" of "Optional[IO[Any]]" has no
attribute "close"  [union-attr]
                    cpio.stdin.close()
                    ^
Found 2 errors in 1 file (checked 3 source files)
Error: Process completed with exit code 1.
```
I don't get this here… Seems bogus.

behrmann commented:
> ```
> Run python3 -m mypy mkosi
> mkosi/__init__.py:3871:21: error: Item "None" of "Optional[IO[Any]]" has no
> attribute "write"  [union-attr]
>                         cpio.stdin.write(file.encode("utf8") + b"\0")
>                         ^
> mkosi/__init__.py:3872:17: error: Item "None" of "Optional[IO[Any]]" has no
> attribute "close"  [union-attr]
>                     cpio.stdin.close()
>                     ^
> Found 2 errors in 1 file (checked 3 source files)
> Error: Process completed with exit code 1.
> ```
> 
> I don't get this here… Seems bogus.

This is semi-bogus, I think. I'm not sure what the idea is, maybe they think, that the subprocess could have closed stdin? Adding a `assert cpio.stdin is not None` below resolves the issue.

behrmann commented:
Ah, I only just saw, that this is a new `popen` function. Yes, I guess is that mypy can't properly determine the type of `stdin` in the Popen and throws this error, because it could be `None`.

keszybz commented:
But I'm passing `stdin=subprocess.PIPE`, so we know that it is not None. Even if the file was closed, the object is not replaced by None, just a call to `.write()` would return an error. I will add the `assert` as suggested if that is necessary to resolve the warning, but I don't think mypy is being smart here. (I'll wait to see if there are any other comments though. I don't think it's worth to restart the CI just for that.)

behrmann commented:
That's true and it's clearly a bug. Would you care to open on against [mypy](https://github.com/python/mypy/issues)?

You can also add an appropriate type ignore comment to mollify mypy.

keszybz commented:
https://github.com/python/mypy/issues/10583

keszybz commented:
Now with a workaround for mypy.

I was experimenting with redirecting stdin from /dev/null for subprocesses. It doesn't help with the mypy error, but I think it's good to do that anyway, so I left that as a separate commit.

behrmann reviewed (commented):
Only did a quick readthrough and haven't tested this yet. I like the new option names a lot better. 

I think it would make sense to also add some test for the cpio output, so that we don't break it by chance at some future point.

poettering commented:
> So... let's say we add EROFS support. (I think this could be a nicer alternative to squashfs images.)
> EROFS has inner compression (lz4 only?). It might be useful to add outer compression with `xz` or `zstd` to shave off some bytes for distribution. It doesn't fit to call the outer compression "archive" in this case, because there's not archive involved.

The gzip manpage uses the word "archive" when talking about anything *.gz.  The "xz" manpaged uses it generically too to some point, but less so. (zstd man page doesn't mention the word at all). Hence I think it would not be too far off to say that a disk image that is compressed with one of the three compressors at the end, without a tar envelope is still an archive.

poettering commented on mkosi/__init__.py:
```diff
@@ -3725,20 +3764,61 @@ def make_tar(args: CommandLineArguments, root: str, do_run_build_script: bool, f
         f: BinaryIO = cast(BinaryIO, tempfile.NamedTemporaryFile(dir=os.path.dirname(args.output), prefix=".mkosi-"))
         # OpenMandriva defaults to bsdtar(libarchive) which uses POSIX argument list so let's keep a separate list
         if shutil.which("bsdtar") and args.distribution == Distribution.openmandriva:
-            _tar_cmd = ["bsdtar", "-C", tar_root_dir, "-c", "-J", "--xattrs", "-f", "-", "."]
```
(btw, i think to keep things systematic and well-defined on all distros we should always use "gtar" here first, and fall back to "tar", but never use bsdtar. In particular as bsdtar has no selinux control iirc)

but that's something to fix separately.

poettering commented on man/mkosi.1:
```diff
@@ -392,33 +393,39 @@ against off-line modification, the verification data is placed in an
-however this option may be used to select the algorithm.
+but this option may be used to select the algorithm.
+.TP
+\f[B]\f[CB]--compress-output=\f[B]\f[R]
```
"output" sounds good to me too, maybe indeed better than "archive"

poettering commented:
looks all good to me, but still no the python guru who could say overly smart things about the pythonicness of the code

keszybz commented on mkosi/__init__.py:
```diff
@@ -251,13 +260,46 @@ def run(
+        die(f"{cmdline[0]} not found in PATH.")
+
+
+def popen(
```
I did it because there's a `run()` function, which invokes `subprocess.run()`. So a new function which invokes `subprocess.Popen()` is called `popen()`. But I don't think the `Popen()` name is very good, so I'm happy to rename this. Maybe `spawn` ?

behrmann replied:
I think I like spawn.

keszybz replied:
I rewrote the two functions a bit: now there's spawn() which does all the work, and run() which is a thin wrapper. This way it's not necessary to duplicate the logging and argument mangling logic.

keszybz replied:
I rewrote the wrapper again. I had trouble getting all the stdin/stdout/input/output/stderr/text cases right.

keszybz commented on mkosi/__init__.py:
```diff
@@ -3713,6 +3797,23 @@ def make_read_only(args: CommandLineArguments, root: str, for_cache: bool, b: bo
+
+    if option == "xz":
+        return [xz_binary(), "--check=crc32", "--lzma2=dict=1MiB", "-T0"]
+    elif option == "zstd" or option is True:
```
Indeed, this was a left-over from previous times.

keszybz commented:
Updated with the requested changes.
I didn't add a test though: I'll work in that next.

behrmann commented on mkosi/__init__.py:
```diff
@@ -236,28 +238,71 @@ def handler(signal: int, frame: FrameType) -> None:
-            return subprocess.run(cmdline, check=check, stdout=stdout, stderr=stderr, **kwargs)
+        with cm():
+            return subprocess.run(
+                cmdline, check=check, input=input, stdin=stdin, stdout=stdout, stderr=stderr, **kwargs
```
This can potentially byte a user, since `input` and `stdin` cannot be non-`None` at the same time. It might make more sense to standardise on either in our wrapper.

keszybz replied:
There's a check above, and I think that should be enough. Standardizing on one of them is not easy, because they have completely different types and semantics...

behrmann replied:
True, the thought came from the above check being for both not being `None`, but in the end this will be caught in `subprocess.run` itself, so that's fine.

behrmann commented on mkosi/__init__.py:
```diff
@@ -3721,20 +3825,69 @@ def make_tar(args: CommandLineArguments, root: str, do_run_build_script: bool, f
+        compressor = compressor_command(should_compress_output(args))
+
+        with spawn(cmd, stdout=subprocess.PIPE) as tar:
+            run(compressor, stdin=tar.stdout, stdout=f, delay_interrupt=False)
```
Is there an advantage not to use tar's `--use-compress-program` as `f"--use-compress-program='{compressor}'"`? The short option is different between GNU tar and BSD tar, but the long option has the same name.

keszybz replied:
Indeed. I'll rework it like that.

keszybz commented:
Updated with two changes:
- `tar --use-compress-program=` is used instead of building a pipeline internally
- I dropped the commit to redirect stdin from `/dev/null`. I didn't investigate the details, but I think this was causing the CI failure. I'll submit that separately once dust settles.

keszybz commented:
CI passes, yay. Let's give it more to chew.

keszybz commented:
Rebased for the backend.py split.

behrmann commented:
> Rebased for the backend.py split.

Thanks! I almost feared I would have to pick out the things in #729 again :)

I'll have a look at this PR later today.

behrmann commented on mkosi/__init__.py:
```diff
@@ -5308,8 +5405,10 @@ def xescape(s: str) -> str:
+def build_auxiliary_output_path(args: argparse.Namespace, suffix: str, can_compress: bool = False) -> str:
+    output = strip_suffixes(args.output)
+    compress = should_compress_output(args) if can_compress else False
+    return output + suffix + (f".{compress}" if compress else "")
```
Just a stylistic quibble, feel free to ignore it
```suggestion
    return f"{output}{suffix}.{compress or ''}"
```

keszybz replied:
No, that makes the dot unconditional, but it should be added only if the compression suffix is added.

behrmann replied:
Right, chalk it up to post-vaccine grogginess :)

behrmann reviewed (approved):
LGTM. Haven't run it myself, but there's little I could do that the tests don't already do. 

@DaanDeMeyer Can you have a look at this so that another pair of eyes looked this over?

keszybz commented:
@DaanDeMeyer any comments?

DaanDeMeyer commented:
No time to give this a thorough look! But the changes look good and given @behrmann gave this a thorough look, I think we can merge.

(Given that I'm swamped with other things at the moment, feel free to merge things yourself @keszybz @poettering if I take too long)
